### PR TITLE
fix: errors when building theme using CSS variables

### LINKED ISSUE #15107: Allow CSS variables in color paletts
#### Please describe the feature you would like to request.
A color pallet has to be defined by by using COLOR values, but doesn't allow CSS variables. 

Example:
````
$sample-pallet: (
  50 : var(--color-primary), ...
````
is invalid.

#### What is the use-case or motivation for this proposal?
We have a application, where we defined the colors as CSS variables, so we can switch for  different brandings.

#### Is there anything else we should know?
Only two places are depending on the colors:

```
  .mat-calendar-body-disabled > .mat-calendar-body-selected {
    background-color: fade-out(mat-color($palette), $mat-datepicker-selected-fade-amount);
  }
```
_theming.scss 2542

```
.mat-calendar-body-disabled > .mat-calendar-body-today:not(.mat-calendar-body-selected) {
    border-color: fade-out(mat-color($foreground, hint-text), $mat-datepicker-today-fade-amount);
  }
```
_theming.scss 2603

If these lines are commented out, a `pallet` can be used with `CSS variables`. 

I know CSS variables are not supported by Internet explorer. However just changing these two lines (the calendar disables state), would allow their usage!

### PR DESCRIPTION
**Pull request by crisbeto**
* Adds an extra theme to the build that will allow us to verify that everything still works if somebody passes in the palette colors as CSS variables.
* Fixes a handful of places where Sass would either throw an error or would produce invalid CSS.

Fixes #15107.

### CODE REVIEW

File: src/lib/core/theming/test-css-variables-theme.scss [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  @return $output;
+}
+
+// Theme used to test that our themes would compile if the colors were specified as CSS variables.
+._demo-css-variables-theme {
+  $palette: mat-palette($mat-blue-grey);
+  $theme: mat-dark-theme($palette, $palette, $palette);
+  @include angular-material-theme(replace-all-values($theme, var(--test-var)));
```
Comment (mmalerba) [inline src/lib/core/theming/test-css-variables-theme.scss]: I'm curious how things would actually look with the changes you've made, could we make a similar function that stores off the current values in the map (e.g. `--test-var-#{n}: $value`) and then updates the map with variables referring to the original value (e.g. `var(--test-var-#{n})`). That way we could compare how the app looks with the original theme and with the variable alias theme

Comment (crisbeto) [reply src/lib/core/theming/test-css-variables-theme.scss]: We should be able to get it to work, but I'd rather get this in first so we don't continue introducing new failures.

File: src/lib/badge/_badge-theme.scss [hunk 2]
```diff
@@ -136,8 +136,15 @@ $mat-badge-large-size: $mat-badge-default-size + 6;
       // percentage of the background to put into foreground when mixing the colors together.
       $app-background: mat-color($background, 'background');
       $badge-color: mat-color($foreground, disabled-button);
-      $badge-opacity: opacity($badge-color);
-      background: mix($app-background, rgba($badge-color, 1), (1 - $badge-opacity) * 100%);
+
+      @if (type-of($badge-color) == color and type-of($app-background) == color) {
```
Comment (jelbourn) [inline src/lib/badge/_badge-theme.scss]: Add a comment along the lines:
```scss
// If the given color is actually a color *type*, we _____. If the given background is not
// specifically a color (e.g. a css variable), use it directly.
```
(similar elsewhere)

File: src/lib/core/theming/test-css-variables-theme.scss (hunk 1 above)
Comment (mmalerba) [reply src/lib/core/theming/test-css-variables-theme.scss]: I suppose it doesn't make things any worse for users

**Pull request merged**